inode2prog.cpp

```diff
@@ -73,46 +79,59 @@ int str2int (const char * ptr) {
 	return retval;
 }
 
-// Not sure, but assuming there's no more PID's than go into 64 unsigned bits..
-#define MAX_PID_LENGTH 20
+static std::string read_file (int fd) {
+	char buf[255];
+	std::string content;
 
-char * getprogname (pid_t pid) {
-	int bufsize = 80;
-	char buffer [bufsize];
+	for (int length; (length = read(fd, buf, sizeof(buf))) > 0;) {
+		if (length < 0) {
+			std::fprintf(stderr, "Error reading file: %s\n", std::strerror(errno));
+			std::exit(34);
+		}
+		content.append(buf, length);
+	}
 
-	int maxfilenamelen = 14 + MAX_PID_LENGTH + 1; 
-	char filename[maxfilenamelen];
+	return content;
```

Comment:
I just realized that there might actually be a away around the extra copy short of C++11's move abilities. I need to look into this more (hopefully I'll have an update tomorrow), but I believe the following avoids the potential copy in the copy assignment provided the compiler supports even basic NRVO:

    std::string tmp = getprogname(pid);
    std::swap(pn->name, tmp);

Basically `tmp` will take over the `content` allocation in `read_file`, then the buffer from `tmp` will get swapped into `pn->name`, and the `pn->name` buffer will get freed up when `tmp` goes out of scope.

This still isn't fully optimal since whatever buffer is in `pn->name` could theoretically get reused instead of thrown away, but the existing code doesn't do this anyway, so this the swap should make it essentially equivalent to the `s->name = strdup()` method, except the memory leak will get cleaned up.